test/libsolidity/smtCheckerTests/function_selector/function_types_sig.sol

```diff
@@ -24,10 +24,8 @@ contract C {
     }
 }
 // ----
-// Warning 6031: (261-267): Internal error: Expression undefined for SMT solver.
 // Warning 7650: (284-296): Assertion checker does not yet support this expression.
 // Warning 6328: (470-495): CHC: Assertion violation happens here.\nCounterexample:\nx = 0\n\nTransaction trace:\nC.constructor()\nState: x = 0\nC.check()\n  C.f() -- internal call\n  C.g() -- internal call
 // Warning 6328: (540-565): CHC: Assertion violation happens here.\nCounterexample:\nx = 0\n\nTransaction trace:\nC.constructor()\nState: x = 0\nC.check()\n  C.f() -- internal call\n  C.g() -- internal call\n  C.i() -- internal call\n  C.i() -- internal call
-// Warning 6031: (261-267): Internal error: Expression undefined for SMT solver.
```

Comment:
I'm wondering why these went away

Reply:
Me too, I will investigate.

Reply:
So, the difference is that before when processing the assignment `function () pure external returns (bytes4) fun = this.f` the SMT expression for the RHS was not created before a call to `SMTEncoder::assignment`, which calls `SMTEncoder::expr`, which in turn reports this error if the expression has not been yet created. But then it creates it and continues as expected.

On the other hand, with the new refactoring, we first call `m_context.expression` on RHS, so the expression will exist on the call to `SMTEncoder::expr` and there is no error to report.

I am wondering if we should define some invariant or rule regarding the creation of expression. Or is there one already and I am not following it?

Reply:
I don't think there is a precise one. Well, ideally we wouldn't have any `Internal error: Expression undefined for SMT solver.`, but this happens currently only for function types I think. So I think this PR makes it better.